File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lrclyrics/db/SearchCache.java
package com.wa2c.android.medoly.plugin.action.lrclyrics.db;

import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.github.gfx.android.orma.annotation.Column;
import com.github.gfx.android.orma.annotation.Index;
import com.github.gfx.android.orma.annotation.PrimaryKey;
import com.github.gfx.android.orma.annotation.Table;
import com.google.gson.Gson;
import com.wa2c.android.medoly.plugin.action.lrclyrics.search.ResultItem;

import java.io.Serializable;
import java.util.Date;

/**
 * Search condition. (Orma Model)
 */
@Table(value = "search_cache", indexes = @Index(value= {"title", "artist"}))
public class SearchCache implements Serializable {

    /** Key. */
    @PrimaryKey(autoincrement = true)
    public long _id;

    /** Search title. */
    @Column(indexed = true)
    public String title;

    /** Search artist. */
    @Nullable
    @Column(indexed = true)
    public String artist;

    /** Search language. */
    @Nullable
    @Column
    public String language;

    /** Search result uploader. */
    @Nullable
    @Column
    public String from;

    /** Search result file name. */
    @Nullable
    @Column
    public String file_name;

    /** Search result has lyrics. */
    @Nullable
    @Column
    public Boolean has_lyrics;

    /** Search result. (JSON)  */
    @Column
    public String result;

    /** Added date. */
    @Nullable
    @Column(defaultExpr = "CURRENT_TIMESTAMP")
    public Date date_added;

    /** Modified date. */
    @Nullable
    @Column(defaultExpr = "CURRENT_TIMESTAMP")
    public Date date_modified;




    /**
     * Get ResultItem from result field.
     * @return result item.
     */
    public ResultItem makeResultItem() {
        if (result == null || result.isEmpty())
            return null;
        return (new Gson()).fromJson(result, ResultItem.class);
    }

    /**
     * Set ResultItem to result field.
     * @param item result item.
     */
    public void updateResultItem(ResultItem item) {
        if (item == null)
            return;
        result = (new Gson()).toJson(item);
    }

}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lrclyrics/db/SearchCache.kt
package com.wa2c.android.medoly.plugin.action.lrclyrics.db

import com.github.gfx.android.orma.annotation.*
import com.google.gson.Gson
import com.wa2c.android.medoly.plugin.action.lrclyrics.search.ResultItem
import java.io.Serializable
import java.util.*

/**
 * Search condition. (Orma Model)
 */
@Table(value = "search_cache", indexes = [Index(value = ["title", "artist"])])
data class SearchCache (
        @Setter @PrimaryKey(autoincrement = true) var _id: Long,
        @Setter @Column(indexed = true) var title: String?,
        @Setter @Column(indexed = true) var artist: String?,
        @Setter @Column var language: String?,
        @Setter
        @Column var from: String?,
        @Setter @Column var file_name: String?,
        @Setter @Column var has_lyrics: Boolean?,
        @Setter @Column var result: String?,
        @Setter @Column(defaultExpr = "CURRENT_TIMESTAMP") var date_added: Date?,
        @Setter @Column(defaultExpr = "CURRENT_TIMESTAMP") var date_modified: Date?
): Serializable {

    /**
     * Get ResultItem from result field.
     * @return result item.
     */
    fun makeResultItem(): ResultItem? {
        return if (result == null || result!!.isEmpty()) null else Gson().fromJson(result, ResultItem::class.java)
    }

    /**
     * Set ResultItem to result field.
     * @param item result item.
     */
    fun updateResultItem(item: ResultItem?) {
        if (item == null)
            return
        result = Gson().toJson(item)
    }

}
